Eliminate the EventMachine dependency

matthewd opened the pull request:
@pixeltrix pointed out that it wasn't actually all that infeasible for us to just import the portion of faye-websocket we need... it turns out to only be about 200 lines. So I did that.

All the protocol complexities live in websocket-driver, which we're still using, so we're not taking on any substantial maintenance burden.

For this relatively low cost, I'd contemplate this just to reduce our dependency list by the one gem... but the important part here is that by importing it, we're able to (as I have) then modify it to avoid EventMachine.

matthewd committed:
Using a hacked faye-websocket, drop EventMachine

matthewd committed:
Import the relevant portions of faye-websocket

(as adapted to use concurrent-ruby / nio4r instead of eventmachine)

schneems commented:
👏

pixeltrix commented on actioncable/lib/action_cable/connection/client_socket.rb:
```diff
@@ -0,0 +1,177 @@
+        return true if env['rack.url_scheme'] == 'https'
+
+        return false
+      end
```
We can utilise `Rack::Request#ssl?` for this can't we?

matthewd replied:
Indeed.. I'll look at chasing `env` -> `req` right through in a subsequent change, because I imagine that'll make @tenderlove happy.

pixeltrix commented on actioncable/lib/action_cable/connection/stream_pool.rb:
```diff
@@ -0,0 +1,73 @@
+
+module ActionCable
+  module Connection
+    class StreamPool
```
Is this a pool or more of a buffer?

matthewd replied:
It doesn't really feel like a buffer either... renamed to StreamEventLoop.

pixeltrix commented on actioncable/lib/action_cable/connection/stream_pool.rb:
```diff
@@ -0,0 +1,73 @@
+require 'concurrent'
```
Are we actually using any concurrent code in this file?

matthewd replied:
Oops.. I deleted the thing that was using it :+1:

pixeltrix commented on actioncable/lib/action_cable/process/logging.rb:
```diff
@@ -1,7 +1,8 @@
+  EventMachine.error_handler do |e|
+    puts "Error raised inside the event loop: #{e.message}"
+    puts e.backtrace.join("\n")
+  end
```
I assume this is still here for when you're using the em-hiredis backend?

matthewd replied:
Yeah. I think I'll just kill it; we're not doing "real work" in the EM loop any more, so it's unlikely to be interesting.

pixeltrix commented on actioncable/test/test_helper.rb:
```diff
@@ -13,24 +13,16 @@
-end
 
-class ActionCable::TestCase < ActiveSupport::TestCase
   def run_in_eventmachine
```
Probably want to rename this method eventually :smile:

pixeltrix commented:
@matthewd :+1: - does this mean we could run it all in one server in development mode?

kddeisz commented:
@matthewd :+1: Does this mean there's work to be done on reloading?

matthewd committed:
Ditch the EM error logging helper

We're no longer doing our work in the EM event loop, so errors are quite
unlikely, and if they do occur, they're not really our responsibility to
handle.

egilburg commented on actioncable/lib/action_cable/connection/client_socket.rb:
```diff
@@ -0,0 +1,177 @@
+          headers.each { |name, value| @driver.set_header(name, value) }
+        end
+
+        [*options[:extensions]].each do |extension|
```
why use two different array handling styles between `:headers` and `:extensions`?

kaspth commented on actioncable/lib/action_cable/connection/client_socket.rb:
```diff
@@ -0,0 +1,177 @@
+        if callback = @env['async.callback']
+          callback.call([101, {}, @stream])
+        end
+      end
```
As far as I can see we aren't passing in protocols or options, why copy them over?

matthewd replied:
Good point. Gone.

maclover7 commented on actioncable/lib/action_cable/connection/client_socket.rb:
```diff
@@ -0,0 +1,155 @@
+        [ -1, {}, [] ]
+      end
+
+
```
:scissors: extra space

maclover7 commented on actioncable/lib/action_cable/connection/client_socket.rb:
```diff
@@ -0,0 +1,155 @@
+      CLOSING    = 2
+      CLOSED     = 3
+
+
```
:scissors: 2 extra spaces

mperham commented on actioncable/lib/action_cable/connection/stream_event_loop.rb:
```diff
@@ -0,0 +1,68 @@
+        @todo = Queue.new
+
+        Thread.new do
+          Thread.current.abort_on_exception = true
```
You probably don't want this.

dhh commented:
Testing this with Basecamp, we're getting two separate exceptions. First:

```
/u/apps/bc3/shared/bundle-ruby221/ruby/2.2.0/bundler/gems/rails-755f7b58953c/actioncable/lib/action_cable/connection/stream_event_loop.rb:56:in `read_nonblock': Connection reset by peer (Errno::ECONNRESET)
from /gems/rails-755f7b58953c/actioncable/lib/action_cable/connection/stream_event_loop.rb:56:in `block (2 levels) in run'
from /gems/rails-755f7b58953c/actioncable/lib/action_cable/connection/stream_event_loop.rb:51:in `each'
from /gems/rails-755f7b58953c/actioncable/lib/action_cable/connection/stream_event_loop.rb:51:in `block in run'
from /gems/rails-755f7b58953c/actioncable/lib/action_cable/connection/stream_event_loop.rb:40:in `loop'
from /gems/rails-755f7b58953c/actioncable/lib/action_cable/connection/stream_event_loop.rb:40:in `run'
from /gems/rails-755f7b58953c/actioncable/lib/action_cable/connection/stream_event_loop.rb:14:in `block in initialize'
```

Which then leads to:

```
/u/apps/bc3/shared/bundle-ruby221/ruby/2.2.0/bundler/gems/rails-755f7b58953c/actioncable/lib/action_cable/connection/base.rb:132:in `on_close': wrong number of arguments (2 for 0) (ArgumentError)
from /gems/rails-755f7b58953c/actioncable/lib/action_cable/connection/client_socket.rb:148:in `finalize_close'
from /gems/rails-755f7b58953c/actioncable/lib/action_cable/connection/client_socket.rb:102:in `client_gone'
from /gems/rails-755f7b58953c/actioncable/lib/action_cable/connection/stream.rb:24:in `close'
from /gems/rails-755f7b58953c/actioncable/lib/action_cable/connection/stream_event_loop.rb:60:in `rescue in block (2 levels) in run'
from /gems/rails-755f7b58953c/actioncable/lib/action_cable/connection/stream_event_loop.rb:55:in `block (2 levels) in run'
from /gems/rails-755f7b58953c/actioncable/lib/action_cable/connection/stream_event_loop.rb:51:in `each'
from /gems/rails-755f7b58953c/actioncable/lib/action_cable/connection/stream_event_loop.rb:51:in `block in run'
from /gems/rails-755f7b58953c/actioncable/lib/action_cable/connection/stream_event_loop.rb:40:in `loop'
from /gems/rails-755f7b58953c/actioncable/lib/action_cable/connection/stream_event_loop.rb:40:in `run'
from /gems/rails-755f7b58953c/actioncable/lib/action_cable/connection/stream_event_loop.rb:14:in `block in initialize'
```
